external/esp_idf_port/esp_lwip/dhcpserver.c

```diff
@@ -0,0 +1,1362 @@
+ * FunctionName : add_end
+ * Description  : add end option of DHCP message
+ * Parameters   : optptr -- the addr of DHCP message option
+ * Returns      : the addr of DHCP message option
+*******************************************************************************/
+static u8_t *add_end(u8_t *optptr)
+{
+	*optptr++ = DHCP_OPTION_END;
+	return optptr;
+}
+
+/******************************************************************************
+ * FunctionName : create_msg
+ * Description  : create response message
+ * Parameters   : m -- DHCP message info
+ * Returns      : none
+*******************************************************************************/
+static void create_msg(struct dhcps_msg *m)
+{
+	ip4_addr_t client;
+
+	client.addr = *((uint32_t *)&client_address);
+
+	m->op = DHCP_REPLY;
+
+	m->htype = DHCP_HTYPE_ETHERNET;
+
+	m->hlen = 6;
+
+	m->hops = 0;
+//        os_memcpy((char *) xid, (char *) m->xid, sizeof(m->xid));
+	m->secs = 0;
+	m->flags = htons(BOOTP_BROADCAST);
+
+	memcpy((char *)m->yiaddr, (char *)&client.addr, sizeof(m->yiaddr));
+
+	memset((char *)m->ciaddr, 0, sizeof(m->ciaddr));
+
+	memset((char *)m->siaddr, 0, sizeof(m->siaddr));
+
+	memset((char *)m->giaddr, 0, sizeof(m->giaddr));
+
+	memset((char *)m->sname, 0, sizeof(m->sname));
+
+	memset((char *)m->file, 0, sizeof(m->file));
+
+	memset((char *)m->options, 0, sizeof(m->options));
+
+	u32_t magic_cookie_temp = magic_cookie;
+
+	memcpy((char *)m->options, &magic_cookie_temp, sizeof(magic_cookie_temp));
+}
+
+struct pbuf *dhcps_pbuf_alloc(u16_t len)
+{
+	u16_t mlen = sizeof(struct dhcps_msg);
+
+	if (len > mlen) {
+#if DHCPS_DEBUG
+		DHCPS_LOG("dhcps: len=%d mlen=%d", len, mlen);
+#endif
+		mlen = len;
+	}
+
+	return pbuf_alloc(PBUF_TRANSPORT, mlen, PBUF_RAM);
+}
+
+/******************************************************************************
+ * FunctionName : send_offer
+ * Description  : DHCP message OFFER Response
+ * Parameters   : m -- DHCP message info
+ * Returns      : none
+*******************************************************************************/
+static void send_offer(struct dhcps_msg *m, u16_t len)
+{
+	u8_t *end;
+	struct pbuf *p, *q;
+	u8_t *data;
+	u16_t cnt = 0;
+	u16_t i;
+	err_t SendOffer_err_t;
+	create_msg(m);
+
+	end = add_msg_type(&m->options[4], DHCPOFFER);
+	end = add_offer_options(end);
+	end = add_end(end);
+
+	p = dhcps_pbuf_alloc(len);
+#if DHCPS_DEBUG
+	DHCPS_LOG("udhcp: send_offer>>p->ref = %d\n", p->ref);
+#endif
+
+	if (p != NULL) {
+
+#if DHCPS_DEBUG
+		DHCPS_LOG("dhcps: send_offer>>pbuf_alloc succeed\n");
+		DHCPS_LOG("dhcps: send_offer>>p->tot_len = %d\n", p->tot_len);
+		DHCPS_LOG("dhcps: send_offer>>p->len = %d\n", p->len);
+#endif
+		q = p;
+
+		while (q != NULL) {
+			data = (u8_t *) q->payload;
+
+			for (i = 0; i < q->len; i++) {
+				data[i] = ((u8_t *) m)[cnt++];
+#if DHCPS_DEBUG
+				DHCPS_LOG("%02x ", data[i]);
+
+				if ((i + 1) % 16 == 0) {
+					DHCPS_LOG("\n");
+				}
+#endif
+			}
+
+			q = q->next;
```

Comment:
Why there are too much empty lines in this file?

Reply:
done